polyhedral: more standard constructions

micjoswig opened the pull request:
Many new constructions of polytopes were added.

TODO:
- add more
- defer long_and_winding (because of coordinate type)
- adjust names, i.e., add "polytope" more often

deferred generalized_permutahedron (because of type conversion issue)

zkgeiselmann committed:
associahedron done

zkgeiselmann committed:
sim body und associahedron ok, documentation unfinished

zkgeiselmann committed:
associahedron and SIM_body done including doc

Commit:
binary_markov_graph

Commit:
cyclic_caratheodory

zkgeiselmann committed:
dwarfed_cube and dwarfed_product_polytope

lkastner committed:
PolyhedralGeometry: Fix reference

zkgeiselmann committed:
build_doc succesfull

zkgeiselmann committed:
lecture_hall_simplex ok, explicit_zonotope not yet, documentation fixed

Commit:
cyclic_caratheodory

Commit:
binary_markov_graph fix

Commit:
fractional_knapsack fix

Commit:
hypersimplex added

zkgeiselmann committed:
goldfarb and type fixes in the zonotopes

Commit:
diverge resolved

Commit:
merge resolved in constructions.md

zkgeiselmann committed:
goldfarb, goldfarb_sit, arg error fixes

Commit:
klee_minty_cube & k_cyclic
:q

zkgeiselmann committed:
long_and_winding

Commit:
Merge conflict resolved

Commit:
max_gc_rank,...pile added

micjoswig committed:
merge

micjoswig committed:
doc fix

zkgeiselmann committed:
did what benjamin said

zkgeiselmann committed:
name fixes, polyhedron vs Polyhedron fixes

zkgeiselmann committed:
minor name fixes in docs and error messages

Commit:
pitman_stanley, pseud_del_pezzo

Commit:
rand01 polytopes added

Commit:
rand_box and rand_cyclic added

zkgeiselmann committed:
generalized_permutahedron unfinished

zkgeiselmann committed:
rss_associahedron

zkgeiselmann committed:
signed_permutahedron

Commit:
rand_normal_polytope added

Commit:
merge conflict resolved

zkgeiselmann committed:
fixed undefined exports and docs in operations on polyhedra

zkgeiselmann committed:
more docs and example fixes

zkgeiselmann committed:
Revert "pitman_stanley, pseud_del_pezzo"

This reverts commit 4107f51f88113c3969b873329b979ed72473d79a.

zkgeiselmann committed:
stable_set

zkgeiselmann committed:
Merge branch 'master' into zkg/associahedron

zkgeiselmann committed:
empty string error resolved

zkgeiselmann committed:
branches have diverged, merge

lgoettgens commented on Project.toml:
```diff
@@ -20,6 +20,7 @@ RandomExtensions = "fb686558-2515-59ef-acaa-46db3789a887"
 RecipesBase = "3cdcf5f2-1ef4-517c-9805-6587b60abb01"
 Singular = "bcd08a7b-43d2-5ff7-b6d4-c458787f915c"
 TOPCOM_jll = "36f60fef-b880-50dc-9289-4aaecee93cc3"
+Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
```
```suggestion
```
What is this needed for? For the tests to succeed, it should be enough for it to be contained in `test/Project.toml`.

zkgeiselmann committed:
fix docs for elementary polytope fcts

zkgeiselmann committed:
Update Project.toml

Co-authored-by: Lars Göttgens <lars.goettgens@rwth-aachen.de>

zkgeiselmann committed:
transportation, stable_set

zkgeiselmann committed:
Merge branch 'zkg/associahedron' of github.com:oscar-system/Oscar.jl into zkg/associahedron

zkgeiselmann committed:
improve style and docs up to fractional_knapsack_polytope

zkgeiselmann committed:
improve style and docs up to permutahedron and add first tests

zkgeiselmann committed:
Merge branch 'master' into zkg/associahedron

zkgeiselmann committed:
more improvements, more tests

zkgeiselmann committed:
minor fix in hypersymplex, more tests

lgoettgens commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1233,3 +1205,1108 @@ function rand_subpolytope(P::Polyhedron{T}, n::Int; seed=nothing) where T<:scala
+        if alpha[i]<alpha[i+1]
+            throw(ArgumentError("SIM_body_polytope: input is not descending"))
+        end
+    end
```
```suggestion
    @req n >= 1 "Dimension must be at least 1"
    for i in 1:n-1
        @req alpha[i] >= alpha[i+1] "Input is not descending"
    end
```
In OSCAR style, we would use the `@req` macro instead of throwing errors. Furthermore, the name of the function throwing some error needn't be mentioned in the error message as it is already included in the error stack.
This pattern could be applied to many places in your PR.

zkgeiselmann committed:
more tests

Commit:
throw Argumenterror replaced with @req

zkgeiselmann committed:
more tests and minor adjustments

lgoettgens commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1187,14 +1159,14 @@ function rand_spherical_polytope(d::Int, n::Int; distribution::Symbol=:uniform,
   end
-  opts = Dict{Symbol,Any}( :template_parameters => [type] )
+  opts = Dict{Symbol,Any}( :template_parameters => [type] ) # creating the Optionset, the :template_parameters are for templating functions in C++
   if seed != nothing
```
```suggestion
  if seed !== nothing
```
Please check for `nothing` using `===` and `!==` instead of `==` and `!=`. (Reference: https://docs.julialang.org/en/v1.6/manual/performance-tips/#Checking-for-equality-with-a-singleton)

lgoettgens replied:
There are more places in `standard_constructions.jl`

lkastner replied:
@lgoettgens Why remark this here? There are places all over Oscar. And how would our normal devs know that `nothing` is a singleton and that they should look there? Also why recommend these checks and not use Julia's `isnothing`?

If checks for `nothing` should be done in a certain way, then there needs to be a test for that. We do not have the capacity to micromanage our devs on that level.

zkgeiselmann committed:
Update src/PolyhedralGeometry/Polyhedron/standard_constructions.jl

Co-authored-by: Lars Göttgens <lars.goettgens@rwth-aachen.de>

zkgeiselmann committed:
add more tests

Commit:
3 tests for rand polytopes added

Commit:
merge conflict resolved

lkastner committed:
PolyhedralGeometry: Use isnothing

lkastner committed:
Merge branch 'zkg/associahedron' of github.com:oscar-system/Oscar.jl into zkg/associahedron

zkgeiselmann committed:
last minor doc fixed

lgoettgens commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+function SIM_body_polytope(alpha::Vector) 
+    n = length(alpha)
+    @req n >= 1 "dimension must be at least 1"
+    @req issorted(-alpha) "input is not descending"
```
```suggestion
    @req issorted(alpha; rev=true) "input is not descending"
```
Doesn't allocate a new vector

zkgeiselmann committed:
fix name for rand_metric, rand_metric_int

lgoettgens commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
````diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+-x₃ ≦ -1
+```
+"""
+associahedron(d::Int) = Polyhedron{QQFieldElem}(Polymake.polytope.associahedron(d)) 
````
This should catch negative inputs, as the polymake error is not helpful.

lgoettgens commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+function cyclic_caratheodory_polytope(d::Int, n::Int)
+    @req d < n "Dimension d has to be smaller than number of points n."
+    @req d >= 2 "Dimension has to be greater or equal to 2."
+    @req mod(d,2) == 0 "Dimension has to be even."
```
```suggestion
    @req iseven(d) "Dimension has to be even."
```

benlorenz commented on Project.toml:
```diff
@@ -18,6 +18,7 @@ Preferences = "21216c6a-2e73-6563-6e65-726566657250"
 Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 RandomExtensions = "fb686558-2515-59ef-acaa-46db3789a887"
 RecipesBase = "3cdcf5f2-1ef4-517c-9805-6587b60abb01"
+Revise = "295af30f-e4ad-537b-8983-00126c2a3abe"
```
```suggestion
```

benlorenz commented on docs/src/PolyhedralGeometry/Polyhedra/constructions.md:
```diff
@@ -136,19 +136,54 @@ regular_600_cell
+permutahedron
+pile_polytope
+pitman_stanley_polytope
+perles_irrational_8_polytope
```
Why was `perles_nonrational_8_polytope` renamed?

benlorenz commented on docs/src/PolyhedralGeometry/Polyhedra/constructions.md:
```diff
@@ -136,19 +136,54 @@ regular_600_cell
+perles_irrational_8_polytope
+pseudo_del_pezzo_polytope
+rand01_polytope
+rand_box
```
```suggestion
rand_box_polytope
```
I think this would be better?

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1061,6 +1043,12 @@ end
+- `d::Int`: the dimension.
+
+Note that in polymake, this function has an optional Boolean parameter `group`, to
+also construct the symmetry group of the simplex.  
```
```suggestion
```
I don't think it makes sense to mention this here.

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1187,14 +1159,14 @@ function rand_spherical_polytope(d::Int, n::Int; distribution::Symbol=:uniform,
     opts[:precision] = convert(Int64, precision)
   end
-  pm_obj = Polymake.call_function(:polytope, :rand_sphere, d, n; opts...)::Polymake.BigObject
+  pm_obj = Polymake.call_function(:polytope, :rand_sphere, d, n; opts...)::Polymake.BigObject # specifying the Type so it will throw an error if its not this type
```
```suggestion
  pm_obj = Polymake.call_function(:polytope, :rand_sphere, d, n; opts...)::Polymake.BigObject
```
This is to tell the compiler the return type since `call_function` is not type-stable.

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
-  if !is_bounded(P)
-    throw(ArgumentError("rand_subpolytope: Polyhedron unbounded"))
-  end
+  @req is_bounded(P) "Polyhedron is unbounded, has to be bounded"
```
```suggestion
  @req is_bounded(P) "Polyhedron has to be bounded"
```

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
````diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+```
+"""
+function dwarfed_cube(d::Int)
+    @req d>=2 "dwarfed_cube: d >= 2 required"
````
```suggestion
    @req d>=2 "dimension must be at least 2"
```

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+"""
+function dwarfed_product_polygons(d::Int, s::Int) 
+    @req d >= 4 && d%2 == 0 "dwarfed_product_polygons: d >= 4 and even required"
+    @req s>=3 "dwarfed_product_polygons: s >= 3 required"
```
```suggestion
    @req d >= 4 && d%2 == 0 "d must be at least 4 and even"
    @req s>=3 "s must be at least 3"
```

lgoettgens replied:
maybe even
```suggestion
    @req d >= 4 && iseven(d) "d must be at least 4 and even"
    @req s>=3 "s must be at least 3"
```

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
````diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+```
+"""
+function hypersimplex(k::Int, d::Int; no_vertices::Bool=false, no_facets::Bool=false, no_vif::Bool=false) 
+    @req 0 < k < d "hypersimplex: 0 < k < d required"
````
```suggestion
    @req 0 < k < d "0 < k < d required"
```

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
````diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+ [2, 1]
+```
+"""
+zonotope(M::Matrix{<:Number}; rows_are_points::Bool=true, centered::Bool=true)  = Polyhedron{QQFieldElem}(Polymake.polytope.zonotope(Matrix{Polymake.Rational}(M), rows_are_points=rows_are_points, centered=centered)) 
````
````suggestion
    zonotope(M::Matrix{<:Number}; centered::Bool=true)

Create a zonotope from a matrix whose rows are input points.

# Optional Arguments
- `centered::Bool`: This is `true` if the output should be centered; the default is `true`.

# Examples
The following produces a parallelogram with the origin as its vertex barycenter: 
```jldoctest
julia> Z = zonotope([1 0; 1 1])
Polyhedron in ambient dimension 2

julia> vertices(Z)
4-element SubObjectIterator{PointVector{QQFieldElem}}:
 [-1, -1//2]
 [0, -1//2]
 [0, 1//2]
 [1, 1//2]
```
The following produces a parallelogram with the origin being a vertex (not centered case): 
```jldoctest
julia> Z = zonotope([1 0; 1 1], centered = false)
Polyhedron in ambient dimension 2

julia> vertices(Z)
4-element SubObjectIterator{PointVector{QQFieldElem}}:
 [0, 0]
 [1, 0]
 [1, 1]
 [2, 1]
```
"""
zonotope(M::Matrix{<:Number}; centered::Bool=true)  = Polyhedron{QQFieldElem}(Polymake.polytope.zonotope(Matrix{Polymake.Rational}(M), rows_are_points=false, centered=centered)) 
````

I think we should remove the `rows_are_points` flag, this doesn't really make sense for the coordinates we are using in Oscar, just keep it at `false` and polymake will add the homogenization coordinate.

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
````diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+Polyhedron in ambient dimension 2
+```
+"""
+function n_gon(n::Int; r::Real=1, alpha_0::Real=0)
````
```suggestion
function n_gon(n::Int; r::RationalUnion=1, alpha_0::RationalUnion=0)
```

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+end
+
+@doc raw"""
+    n_gon(n::Int; r::Rational=1, alpha_0::Rational=0)
```
```suggestion
    n_gon(n::Int; r::RationalUnion=1, alpha_0::RationalUnion=0)
```

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+    @req length(y) >= 1 "length of input must be at least 1"
+    return polyhedron(Polymake.polytope.pitman_stanley(y))
+end
+pitman_stanley_polytope(y::Vector{Int}) = pitman_stanley_polytope(convert(Vector{Rational}, y))
```
```suggestion
function pitman_stanley_polytope(y::AbstractVector{<:RationalUnion})
    @req length(y) >= 1 "length of input must be at least 1"
    return polyhedron(Polymake.polytope.pitman_stanley(y))
end
pitman_stanley_polytope(y::AbstractVector{<:IntegerUnion}) = pitman_stanley_polytope(QQ.(y))
```

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+        pm_obj = Polymake.call_function(:polytope, :rand01, d, n)::Polymake.BigObject
+    else
+        seed = convert(Int64, seed)
+        #opts = Dict{Symbol,Any}(:seed => seed) #opts macht also nur sinn für meherer eintraege
```
```suggestion
```

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+    else
+        seed = convert(Int64, seed)
+        #opts = Dict{Symbol,Any}(:seed => seed) #opts macht also nur sinn für meherer eintraege
+        pm_obj = Polymake.call_function(:polytope, :rand01, d, n; :seed => seed)::Polymake.BigObject # specifying the Type so it will throw an error if its not this type
```
```suggestion
        pm_obj = Polymake.call_function(:polytope, :rand01, d, n; :seed => seed)::Polymake.BigObject
```

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+end
+
+@doc raw"""
+    rand_box(d::Int, n::Int, b::Int, seed::Int=nothing)
```
```suggestion
    rand_box_polytope(d::Int, n::Int, b::Int, seed::Int=nothing)
```

see earlier

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
````diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+
+# Example
+```jldoctest
+julia> r = rand_box(3,10,3, seed=1)
````
```suggestion
julia> r = rand_box_polytope(3, 10, 3, seed=1)
```

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
````diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+ [0, 1, 2]
+```
+"""
+function rand_box(d::Int, n::Int, b::Int; seed=nothing)
````
```suggestion
function rand_box_polytope(d::Int, n::Int, b::Int; seed=nothing)
```

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+        opts = Dict{Symbol, Int}(:seed => seed)
+        pm_obj = Polymake.call_function(:polytope, :rand_metric, n; opts...)
+    end
+    return matrix(QQ, pm_obj) #statt Matrix{QQFieldElem}(pm_obj) (einheitliches Format in Oskar)
```
```suggestion
    return matrix(QQ, pm_obj)
```

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+3×3 Matrix{Rational}:
+               0//1                …  371474612593257//281474976710656
+ 260222460282405//140737488355328     388326899436839//281474976710656
+ 371474612593257//281474976710656                   0//1
```
```suggestion
[                               0   260222460282405//140737488355328   371474612593257//281474976710656]
[260222460282405//140737488355328                                  0   388326899436839//281474976710656]
[371474612593257//281474976710656   388326899436839//281474976710656                                  0]

```

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+
+Produce a `n`-point metric with random integral distances. 
+The values are uniformily distributed in $\[1,2\]$. The distances are integers and lie in 
+$[10^digits, 10^(digits+1)[$  
```
```suggestion
The values are uniformily distributed in $\[1,2\]$. The distances are integers and lie in 
$[10^digits, 10^(digits+1)[$.
```

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+        opts = Dict{Symbol, Int}(:seed => seed)
+        pm_obj = Polymake.call_function(:polytope, :rand_metric_int, n, digits; opts...)
+    end
+    return Matrix{Int}(pm_obj)
```
```suggestion
    return matrix(ZZ, pm_obj)
```

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+    end
+    return Matrix{Int}(pm_obj)
+end
+
```
```suggestion
```

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+   rss_associahedron(n::Int)
+
+Produce a polytope of constrained expansions in ambient dimension `n` according to 
+[RSS03](@cite)
```
```suggestion
Produce a polytope of constrained expansions in ambient dimension `n` according to [RSS03](@cite).
```

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+0
+
+julia> add_edge!(G, 2, 3)
+0
```
```suggestion
julia> add_edge!(G, 1, 2);

julia> add_edge!(G, 1, 3);

julia> add_edge!(G, 2, 3);
```
Maybe we should remove the return value from `add_edge!`.

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
````diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+[1    1    1//2]
+```
+"""
+zonotope_vertices_fukuda_matrix(M::Matrix{<:Number}, rows_are_points::Bool=true)  = matrix(QQ, Polymake.polytope.zonotope_vertices_fukuda(Matrix{Polymake.Rational}(M), rows_are_points=rows_are_points))
````
```suggestion
zonotope_vertices_fukuda_matrix(M::Union{MatElem, AbstractMatrix}, rows_are_points::Bool=true)  = matrix(QQ, Polymake.polytope.zonotope_vertices_fukuda(Polymake.Matrix{Polymake.Rational}(M), rows_are_points=rows_are_points))
```

benlorenz replied:
Maybe we should remove the `rows_are_points` flag here as well, and then we need to dehomogenize the result.

benlorenz commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
```diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+end
+
+@doc raw"""
+    zonotope_vertices_fukuda(M::Matrix{<:Number}, rows_are_points::Bool=true)
```
```suggestion
    zonotope_vertices_fukuda(M, rows_are_points::Bool=true)
```

benlorenz commented on src/exports.jl:
```diff
@@ -1143,6 +1166,12 @@ export quotient
 export radical_membership
 export rand
+export rand01_polytope
+export rand_box
```
```suggestion
export rand_box_polytope
```

benlorenz commented on test/PolyhedralGeometry/polyhedron.jl:
```diff
@@ -256,7 +256,241 @@ for f in (QQ, ENF)
+                let a = associahedron(4)
+                    @test a isa Polyhedron{T}
+                    @test dim(a) == 4
+                    @test size(Oscar.pm_object(a).FACETS,1) == 14
```
```suggestion
                    @test nfacets(a) == 14
```

benlorenz commented on test/PolyhedralGeometry/polyhedron.jl:
```diff
@@ -256,7 +256,241 @@ for f in (QQ, ENF)
+                let dc = dwarfed_cube(3)
+                    @test dc isa Polyhedron{T}
+                    @test dim(dc) == 3
+                    @test size(Oscar.pm_object(dc).FACETS,1) == 7
```
```suggestion
                    @test nfacets(dc) == 7
```

benlorenz commented on test/PolyhedralGeometry/polyhedron.jl:
```diff
@@ -256,7 +256,241 @@ for f in (QQ, ENF)
+                @test_throws ArgumentError signed_permutahedron(100000)
+                let sph = signed_permutahedron(3) 
+                    @test sph isa Polyhedron{T}
+                    @test size(Oscar.pm_object(sph).VERTICES,1) == 48
```
```suggestion
                    @test nvertices(sph) == 48
```

benlorenz commented on test/PolyhedralGeometry/polyhedron.jl:
```diff
@@ -256,7 +256,241 @@ for f in (QQ, ENF)
+                @test rand_cyclic_polytope(2,4) isa Polyhedron{T}
+                let rcyc = p = rand_cyclic_polytope(3,8, seed = 4)
+                    @test rcyc isa Polyhedron{T}
+                    @test size(Oscar.pm_object(rcyc).VERTICES,1) == 8
```
```suggestion
                    @test nvertices(rcyc) == 8
```

benlorenz commented on test/PolyhedralGeometry/polyhedron.jl:
```diff
@@ -256,7 +256,241 @@ for f in (QQ, ENF)
+                end
+
+                let rmetricint = rand_metric_int(3, 2, seed = 213)
+                    @test rmetricint isa Matrix{Int}
```
```suggestion
                    @test rmetricint isa ZZMatrix
```

benlorenz commented on test/PolyhedralGeometry/polyhedron.jl:
```diff
@@ -256,7 +256,241 @@ for f in (QQ, ENF)
+                @test rand01_polytope(2,4) isa Polyhedron{T}
+                let r_01_p = rand01_polytope(2,4; seed = 47)
+                    @test r_01_p isa Polyhedron{T}
+                    @test Oscar.pm_object(r_01_p).N_VERTICES == 4
```
```suggestion
                    @test nvertices(r_01_p) == 4
```

benlorenz commented on test/PolyhedralGeometry/polyhedron.jl:
```diff
@@ -256,7 +256,241 @@ for f in (QQ, ENF)
+                let pile = pile_polytope([2,2])
+                    @test pile isa Polyhedron{T}
+                    @test ambient_dim(pile) == 3
+                    @test Oscar.pm_object(pile).N_VERTICES == 9
```
```suggestion
                    @test nvertices(pile) == 9
```

benlorenz commented on test/PolyhedralGeometry/polyhedron.jl:
```diff
@@ -256,7 +256,241 @@ for f in (QQ, ENF)
+                @test_throws ArgumentError neighborly_cubical_polytope(3,2)
+                let ncp = neighborly_cubical_polytope(2,3)
+                    @test ncp isa Polyhedron{T}
+                    @test Oscar.pm_object(ncp).N_VERTICES == 8
```
```suggestion
                    @test nvertices(ncp) == 8
```

benlorenz commented on test/PolyhedralGeometry/polyhedron.jl:
```diff
@@ -256,7 +256,241 @@ for f in (QQ, ENF)
+                let mpp = multiplex_polytope(2,3)
+                    @test mpp isa Polyhedron{T}
+                    @test Oscar.pm_object(mpp).N_FACETS == 4
+                    @test Oscar.pm_object(mpp).N_VERTICES == 4
```
```suggestion
                    @test nfacets(mpp) == 4
                    @test nvertices(mpp) == 4
```

lgoettgens commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
````diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+# Example
+The following produces a $3$-dimensional Klee-Minty cube for $e=\frac{1}{3}.
+```jldoctest
+julia> c=goldfarb_cube(3,1//3,0)
````
```suggestion
julia> c = goldfarb_cube(3,1//3,0)
```

lgoettgens commented on src/PolyhedralGeometry/Polyhedron/standard_constructions.jl:
````diff
@@ -1218,18 +1190,1083 @@ julia> nvertices(rand_subpolytope(cube(3), 5))
+    
+# Examples
+```jldoctest
+julia> c=goldfarb_sit_cube(3,1//3,1//2)
````
```suggestion
julia> c = goldfarb_sit_cube(3,1//3,1//2)
```

lkastner committed:
Apply suggestions from code review

Co-authored-by: Benjamin Lorenz <benlorenz@users.noreply.github.com>
Co-authored-by: Lars Göttgens <lars.goettgens@rwth-aachen.de>

lkastner committed:
Merge remote-tracking branch 'origin/master' into zkg/associahedron

lkastner committed:
PolyhedralGeoemtry: perles_irrational back to perles_nonrational

lkastner committed:
PolyhedralGeometry: Remove constructions without coordinates

Remove multiplex_polytope and neighborly_cubical_polytope since these
come without coordinates and lack many other properties as well. Hence
they are almost unusable in Oscar.

lkastner committed:
PolyhedralGeometry: associahedron @req d>=0

Sanitize the input dimension for associahedron as "the polymake error
message is not helpful"

lkastner committed:
PolyhedralGeometry: Fix random doctests

- Remove unseeded tests
- Add seeds
- sort output

lgoettgens commented:
I noticed that parts of the additions are inconsistently formatted. @lkastner could you maybe run the formatter on the large chunks of additions?

lkastner committed:
PolyhedralGeometry: Better input handling and remove nonfuncional flag in zonotope_vertices_fukuda

lkastner committed:
PolyhedralGeometry: Run JuliaFormatter

lkastner committed:
Styleguide: Add paragraph on JuliaFormatter

lkastner commented:
> I noticed that parts of the additions are inconsistently formatted. @lkastner could you maybe run the formatter on the large chunks of additions?

Again, this is an Oscar wide issue, so why remark it here if the rest of the code is not up to this standard? Furthermore this issue could be solved by having appropriate checks placed in the testsuite. Additionally we have nothing in the docs on `JuliaFormatter`, which I assume you meant, so I added a paragraph in the styleguide. Out of curiosity I tried to run JuliaFormatter on the entire codebase and it even throws an error.

I do not think this level of feedback is sustainable. Do you want to remark on these formatting / spacing / whitespace issues on every (many) PR(s) in the future?

lgoettgens commented:
> > I noticed that parts of the additions are inconsistently formatted. @lkastner could you maybe run the formatter on the large chunks of additions?
> 
> Again, this is an Oscar wide issue, so why remark it here if the rest of the code is not up to this standard? Furthermore this issue could be solved by having appropriate checks placed in the testsuite. Additionally we have nothing in the docs on `JuliaFormatter`, which I assume you meant, so I added a paragraph in the styleguide. Out of curiosity I tried to run JuliaFormatter on the entire codebase and it even throws an error.
> 
> I do not think this level of feedback is sustainable. Do you want to remark on these formatting / spacing / whitespace issues on every (many) PR(s) in the future?

If I remember correctly, I discussed that with some people a while ago (cannot remember who) and the conclusion was more or less: We do not run the formatter on everything at some point, as that creates conflicts with all open PRs and makes git blame basically unusable. Instead, new additions should follow the formatting styleguide, and whenever you touch large parts of a function you may format that as well.

However, I do not think this is the right place to discuss it. We should move that to a discussion thread, slack, or a Friday meeting (unfortunately I cannot participate this and next week due to travel), to get a broader range of opinions and then write the result down somewhere, e.g. the dev docs.
